src/openrct2-ui/windows/CustomCurrency.cpp

```diff
@@ -42,204 +42,182 @@ static rct_widget window_custom_currency_widgets[] = {
-{
-    if (dropdownIndex == -1)
-        return;
+            gConfigGeneral.custom_currency_affix = CurrencyDescriptors[EnumValue(CurrencyType::Custom)].affix_unicode;
+            config_save_default();
 
-    if (widgetIndex == WIDX_AFFIX_DROPDOWN_BUTTON)
-    {
-        if (dropdownIndex == 0)
-        {
-            CurrencyDescriptors[EnumValue(CurrencyType::Custom)].affix_ascii = CurrencyAffix::Prefix;
-            CurrencyDescriptors[EnumValue(CurrencyType::Custom)].affix_unicode = CurrencyAffix::Prefix;
-        }
-        else if (dropdownIndex == 1)
-        {
-            CurrencyDescriptors[EnumValue(CurrencyType::Custom)].affix_ascii = CurrencyAffix::Suffix;
-            CurrencyDescriptors[EnumValue(CurrencyType::Custom)].affix_unicode = CurrencyAffix::Suffix;
+            window_invalidate_all();
         }
-
-        gConfigGeneral.custom_currency_affix = CurrencyDescriptors[EnumValue(CurrencyType::Custom)].affix_unicode;
-        config_save_default();
-
-        window_invalidate_all();
     }
-}
-
-static void CustomCurrencyWindowTextInput([[maybe_unused]] struct rct_window* w, rct_widgetindex widgetIndex, char* text)
-{
-    if (text == nullptr)
-        return;
-    int32_t rate;
-    char* end;
-    switch (widgetIndex)
+    void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
     {
-        case WIDX_SYMBOL_TEXT:
-            safe_strcpy(CurrencyDescriptors[EnumValue(CurrencyType::Custom)].symbol_unicode, text, CURRENCY_SYMBOL_MAX_SIZE);
-
-            safe_strcpy(
-                gConfigGeneral.custom_currency_symbol, CurrencyDescriptors[EnumValue(CurrencyType::Custom)].symbol_unicode,
-                CURRENCY_SYMBOL_MAX_SIZE);
+        if (text.empty())
+            return;
+        int32_t rate;
+        char* end;
+        switch (widgetIndex)
+        {
+            case WIDX_SYMBOL_TEXT:
+                safe_strcpy(
+                    CurrencyDescriptors[EnumValue(CurrencyType::Custom)].symbol_unicode, std::string(text).c_str(),
+                    CURRENCY_SYMBOL_MAX_SIZE);
 
-            config_save_default();
-            window_invalidate_all();
-            break;
+                safe_strcpy(
+                    gConfigGeneral.custom_currency_symbol, CurrencyDescriptors[EnumValue(CurrencyType::Custom)].symbol_unicode,
+                    CURRENCY_SYMBOL_MAX_SIZE);
 
-        case WIDX_RATE:
-            rate = strtol(text, &end, 10);
-            if (*end == '\0')
-            {
-                CurrencyDescriptors[EnumValue(CurrencyType::Custom)].rate = rate;
-                gConfigGeneral.custom_currency_rate = CurrencyDescriptors[EnumValue(CurrencyType::Custom)].rate;
                 config_save_default();
                 window_invalidate_all();
-            }
-            break;
+                break;
+
+            case WIDX_RATE:
+                rate = strtol(std::string(text).c_str(), &end, 10);
```

Comment:
Can  you use `String::Parse` in our core/string.hpp file instead of this.

Reply:
I'm having trouble using it, I would love some help.
It expects T to be a number type, but strtol expects char*, so I can't use it to cast text to char*. Am I missing something?

Reply:
you would use it like this: 
```cpp
auto res = String::Parse<int32_t>(text);
if (res.has_value()){
    rate = res.value();
    ...
}```

Reply:
Done.